README.md
# BBox Date Line

[![Build Status](https://travis-ci.org/DenisCarriere/bbox-dateline.svg?branch=master)](https://travis-ci.org/DenisCarriere/bbox-dateline)
[![npm version](https://badge.fury.io/js/bbox-dateline.svg)](https://badge.fury.io/js/bbox-dateline)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/DenisCarriere/bbox-dateline/master/LICENSE)
[![Coverage Status](https://coveralls.io/repos/github/DenisCarriere/bbox-dateline/badge.svg?branch=master)](https://coveralls.io/github/DenisCarriere/bbox-dateline?branch=master)

[![Standard - JavaScript Style Guide](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)

> Modifies a BBox to fit within 180° longitude & 90° latitude.

## Install

```bash
$ yarn add bbox-dateline
```

## Usage

```javascript
const dateline = require('bbox-dateline')
dateline.bbox([190, 100, -200, -120])
//= [-170, -80, 160, 60]
```

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [bbox](#bbox)
-   [center](#center)
-   [latitude](#latitude)
-   [longitude](#longitude)

### bbox

Modifies a BBox to fit within the bounds of the International Date Line.

**Parameters**

-   `bbox` **(BBox | FeatureCollection | Feature&lt;any>)** BBox [west, south, east, north] or GeoJSON Feature

**Examples**

```javascript
dateline.bbox([190, 100, -200, -120])
//= [-170, -80, 160, 60]
```

Returns **BBox** valid BBox extent

### center

Modifies a Center to fit within the bounds of the International Date Line.

**Parameters**

-   `coord` **(\[[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number), [number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)] | BBox | FeatureCollection | Feature&lt;any>)** Center [lng, lat], BBox [west, south, east, south] or GeoJSON Feature

**Examples**

```javascript
dateline.center([190, 100])
//= [-170, -80]
```

Returns **\[[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number), [number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)]** valid center coordinate

### latitude

Modifies a Latitude to fit within +/-90 degrees.

**Parameters**

-   `lat` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** latitude to modify

**Examples**

```javascript
dateline.latitude(100)
//= -80
```

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** modified latitude

### longitude

Modifies a Longitude to fit within +/-180 degrees.

**Parameters**

-   `lng` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** longitude to modify

**Examples**

```javascript
dateline.longitude(190)
//= -170
```

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** modified longitude


CHANGELOG.md
# Changelog

## 1.2.1 - 2017-06-2

- Change input BBox Typescript types to `number[]`

## 1.2.0 - 2017-05-29

- Drop Rollup
- Update typescript definition
- Add Coverage to tests
- Include yarn.lock

## 1.1.0 - 2017-03-20

- Support bbox that overlaps the entire world

## 1.0.0 - 2017-03-17

- Initialize `bbox-dateline` library
- Add methods `bbox()`, `center()`, `latitude()` & `longitude()`

test.js
const {test} = require('tap')
const fs = require('fs')
const path = require('path')
const load = require('load-json-file')
const write = require('write-json-file')
const {lineString} = require('@turf/helpers');
const dateline = require('./')

const directories = {
  in: path.join(__dirname, 'test', 'in') + path.sep,
  out: path.join(__dirname, 'test', 'out') + path.sep
}

const fixtures = fs.readdirSync(directories.in).map(filename => {
  return {
    filename,
    name: path.parse(filename).name,
    geojson: load.sync(directories.in + filename)
  }
})

test('dateline', t => {
  for (const {name, filename, geojson} of fixtures) {
    const bbox = dateline.bbox(geojson)
    geojson.bbox = bbox

    if (process.env.REGEN) write.sync(directories.out + filename, geojson)
    t.deepEqual(geojson, load.sync(directories.out + filename), name)
  }
  t.end()
})

test('center', t => {
  t.deepEqual(dateline.center([190, 100, 210, 120]), [-160, -70])
  t.deepEqual(dateline.center([190, 100]), [-170, -80])
  t.assert(dateline.center(lineString([[0, 0], [10, 10]])))

  t.throws(() => dateline.center([10]), /center requires at least 2 numbers/)
  t.throws(() => dateline.center(), /coord is required/)
  t.end()
})

test('bbox', t => {
  t.deepEqual(dateline.bbox([190, 100, -200, -120]), [-180, -90, 180, 90])
  t.deepEqual(dateline.bbox([-190, 100, 200, -120]), [-180, -90, 180, 90])
  t.deepEqual(dateline.bbox([-190, -100, 200, 120]), [-180, -90, 180, 90])
  t.deepEqual(dateline.bbox([-500, -90, 100, 90]), [-180, -90, 180, 90])
  t.deepEqual(dateline.bbox([100, -90, 500, 90]), [-180, -90, 180, 90])
  t.end()
})

test('latitude', t => {
  t.deepEqual(dateline.latitude(0), 0)
  t.deepEqual(dateline.latitude(-180), 0)
  t.deepEqual(dateline.latitude(80), 80)
  t.deepEqual(dateline.latitude(100), -80)
  t.deepEqual(dateline.latitude(-100), 80)
  t.deepEqual(dateline.latitude(-0), 0)
  t.end()
})

test('longitude', t => {
  t.deepEqual(dateline.longitude(0), 0)
  t.deepEqual(dateline.longitude(-0), 0)
  t.deepEqual(dateline.longitude(-360), 0)
  t.deepEqual(dateline.longitude(160), 160)
  t.deepEqual(dateline.longitude(190), -170)
  t.deepEqual(dateline.longitude(-190), 170)
  t.end()
})


test/types.ts
import {BBox, Feature, Features} from '../'
import * as dateline from '../'

const bbox: BBox = [-180, -90, 180, 90]
const feature: Feature = undefined
const features: Features = undefined

dateline.bbox(bbox)
dateline.bbox(feature)
dateline.center([bbox[0], bbox[1]])
dateline.center(bbox)
dateline.center(feature)
dateline.latitude(bbox[0])
dateline.longitude(bbox[1])


index.js
var turfBBox = require('@turf/bbox')
var turfCenter = require('@turf/center')

/**
 * Modifies a BBox to fit within the bounds of the International Date Line.
 *
 * @param {BBox|FeatureCollection|Feature<any>} bbox BBox [west, south, east, north] or GeoJSON Feature
 * @returns {BBox} valid BBox extent
 * @example
 * dateline.bbox([190, 100, -200, -120])
 * //= [-170, -80, 160, 60]
 */
function bbox (bbox) {
  // input validation
  if (!bbox) throw new Error('bbox is required')
  if (!Array.isArray(bbox)) bbox = turfBBox(bbox)
  if (bbox.length !== 4) throw new Error('bbox must have 4 numbers')

  var west = bbox[0]
  var south = bbox[1]
  var east = bbox[2]
  var north = bbox[3]

  // Support bbox that overlaps the entire world
  if (west < -180 && east > 180) {
    west = -180
    east = 180
  }
  if (east < -180 && west > 180) {
    west = -180
    east = 180
  }
  if (south < -90 && north > 90) {
    south = -90
    north = 90
  }
  if (north < -90 && south > 90) {
    south = -90
    north = 90
  }
  if (north > 90) north = 90
  if (south < -90) south = -90

  // Beyond 360 longitude degrees
  if (Math.abs(bbox[0] - bbox[2]) > 360) {
    west = -180
    east = 180
  }
  // Beyond 180 latitude degrees
  if (Math.abs(bbox[1] - bbox[3]) > 180) {
    south = -90
    north = 90
  }
  // Convert Lat & Lng within dateline
  west = longitude(west)
  south = latitude(south)
  east = longitude(east)
  north = latitude(north)

  return [west, south, east, north]
}

/**
 * Modifies a Center to fit within the bounds of the International Date Line.
 *
 * @param {[number, number]|BBox|FeatureCollection|Feature<any>} coord Center [lng, lat], BBox [west, south, east, south] or GeoJSON Feature
 * @returns {[number, number]} valid center coordinate
 * @example
 * dateline.center([190, 100])
 * //= [-170, -80]
 */
function center (coord) {
  if (!coord) throw new Error('coord is required')

  // Support BBox [west, south, east, north]
  if (Array.isArray(coord)) {
    if (coord.length === 4) {
      var bbox = coord
      var west = bbox[0]
      var south = bbox[1]
      var east = bbox[2]
      var north = bbox[3]
      coord = [(west + east) / 2, (south + north) / 2]
    }
  // Support any GeoJSON
  } else coord = turfCenter(coord).geometry.coordinates

  if (coord.length < 2) throw new Error('center requires at least 2 numbers')
  var lng = longitude(coord[0])
  var lat = latitude(coord[1])

  return [lng, lat]
}

/**
 * Modifies a Latitude to fit within +/-90 degrees.
 *
 * @param {number} lat latitude to modify
 * @returns {number} modified latitude
 * @example
 * dateline.latitude(100)
 * //= -80
 */
function latitude (lat) {
  if (lat === undefined || lat === null) throw new Error('lat is required')

  // Latitudes cannot extends beyond +/-90 degrees
  if (lat > 90 || lat < -90) {
    lat = lat % 180
    if (lat > 90) lat = -180 + lat
    if (lat < -90) lat = 180 + lat
    if (lat === -0) lat = 0
  }
  return lat
}

/**
 * Modifies a Longitude to fit within +/-180 degrees.
 *
 * @param {number} lng longitude to modify
 * @returns {number} modified longitude
 * @example
 * dateline.longitude(190)
 * //= -170
 */
function longitude (lng) {
  if (lng === undefined || lng === undefined) throw new Error('lng is required')

  // lngitudes cannot extends beyond +/-90 degrees
  if (lng > 180 || lng < -180) {
    lng = lng % 360
    if (lng > 180) lng = -360 + lng
    if (lng < -180) lng = 360 + lng
    if (lng === -0) lng = 0
  }
  return lng
}

module.exports = {
  bbox: bbox,
  longitude: longitude,
  latitude: latitude,
  center: center
}
